feat(popover): gen2 popover migration (#6356)
adobe/spectrum-web-components: 2nd-gen/packages/swc/components/popover/popover.mdx

import { Canvas, Meta } from '@storybook/addon-docs/blocks';
import { DocsFooter, DocsHeader } from '../../.storybook/blocks';

import * as Stories from './stories/popover.stories';

<Meta of={Stories} />

<DocsHeader />

## Anatomy

A popover is made up of:

1. **Surface**: the anchored container that holds the content.

> 2. **Default slot**: free-form slotted content such as text, a menu, a form, or an account card.

Reviewer: Let's remove "menu" for now since we want to ultimately guide that experience a bit more, and also not mix in "card" language.

```suggestion
2. **Default slot**: free-form slotted content such as text, a form, or an account profile.
```

Author: Done. The slot now reads "text, a form, or an account profile" (dropped "menu" and "card").

3. **Arrow (tip)**: a triangle pointing back at the trigger, shown by default and removed with `hide-arrow`.

The popover positions itself against a trigger you associate with `for` or `triggerElement`; it does not render the trigger itself.

<Canvas of={Stories.Anatomy} />

## Options

### Placement

> `placement` sets the preferred side relative to the trigger: `top`, `bottom` (default), and the logical inline sides `start` and `end`. Logical sides resolve against the writing direction, so `start` is the left in LTR and the right in RTL (and `end` is the mirror), keeping placements correct under right-to-left layouts. Each side also accepts an alignment toward one end of that edge: `top-start` / `top-end`, `bottom-start` / `bottom-end`, `start-top` / `start-bottom`, and `end-top` / `end-bottom`. When the preferred side lacks room the popover flips to the opposite side, unless `should-flip` is `false`. `offset` (main axis) and `cross-offset` (cross axis) nudge it in pixels from the trigger.

Reviewer: Consider borrowing language from Tooltip to better explain the behavior here, especially around flips.

Author: Done. Reworked Placement to borrow Tooltip's wording: flip behavior, the requested value not being rewritten by the flip, the viewport flip boundary in the top layer, and `actual-placement`.

<Canvas of={Stories.Placement} />

### Sizes

`size` fixes the inline size: `s` (336px), `m` (416px), or `l` (576px). When unset, the popover fits its content up to the viewport width minus the container padding.

<Canvas of={Stories.Sizes} />

### Hide arrow

The arrow (tip) points at the trigger and is shown by default. Set `hide-arrow` to remove it; the surface then uses a rectangular box-shadow, so nested popovers do not show overlapping arrows.

<Canvas of={Stories.HideArrow} />

## States

> ### States

Reviewer 2: This is a duplicate heading.

```suggestion
```

Author: Done. Removed the duplicate `### States` heading.

> `open` is the single source of truth for visibility: the popover is either open or closed. Toggling the trigger, pressing `Escape`, or clicking outside (default mode) all reconcile back to `open`. There are no disabled or selected states.

Reviewer: Consider borrowing Tooltip's language to explain the "auto" behavior closing other popovers, including Tooltips.

Author: Done!

<Canvas of={Stories.States} />

## Behaviors

### Custom anchor

The toggling control and the positioning anchor can be different elements: a button drives `open` while the popover anchors to another element via the `triggerElement` property (`manual` suppresses the anchor's own click-to-toggle).

<Canvas of={Stories.CustomAnchor} />

> 

Reviewer: Possibly include a separate section for Trigger element, like Tooltip?

Author: Done. Consolidated into a single `### Trigger element` section (borrowing Tooltip), folding in the former "Custom anchor" content, and renamed the `CustomAnchor` story to `TriggerElement`.

### Virtual anchor

`triggerElement` also accepts a virtual element: any object with a `getBoundingClientRect()` method, the same shape the `PlacementController` supports. The popover then anchors to an arbitrary point or rect with no backing DOM element, for cases like a context menu at the pointer or a popover anchored to a text selection. A virtual anchor positions only; drive `open` yourself and place ARIA and focus on the real control that invoked it (or pair it with `for` to keep `aria-expanded` / `aria-controls` and focus restoration on a button).

<Canvas of={Stories.VirtualAnchor} />

### Modal

Setting `modal` switches the surface to a `<dialog>` opened with `showModal()`: focus is trapped inside, the page behind is inert and its scroll is locked, and the surface exposes `role="dialog"`. Dismiss with `Escape` or a click on the backdrop. Name the dialog with `accessible-label` (see [Accessibility](#accessibility)).

<Canvas of={Stories.Modal} />

#### Default-mode dismissal

> In the default (non-modal) mode the surface is a `<div popover="auto">`: the browser dismisses it on `Escape` or a click outside its own content, and the page behind stays interactive and scrollable. Clicking or focusing content inside the popover does not dismiss it.

Reviewer 2: ```suggestion
### Focus trapping and dismissal

#### Default-mode dismissal

In the default (non-modal) mode the surface is a `<div popover="auto">`: the browser dismisses it on `Escape` or a click outside its own content, and the page behind stays interactive and scrollable. Clicking or focusing content inside the popover does not dismiss it.

#### Modal

Setting `modal` switches the surface to a `<dialog>` opened with `showModal()`: focus is trapped inside, the page behind is inert and its scroll is locked, and the surface exposes `role="dialog"`. Dismiss with `Escape` or a click on the backdrop. Name the dialog with `accessible-label` (see [Accessibility](#accessibility)).

<Canvas of={Stories.Modal} />
```

Author: Done. Applied your structure: `### Focus trapping and dismissal` with `#### Default-mode dismissal` and `#### Modal` subsections.

#### Focus and events

> Opening the popover moves focus into it and closing returns focus to the trigger. The popover dispatches `swc-open` / `swc-after-open` on open and `swc-close` / `swc-after-close` on close; `swc-close.detail.source` reports why it closed: `escape`, `outside`, or `programmatic`.

Reviewer: Consider expanding the event descriptions, similar to Tooltip.

Author: Done. Expanded Events into a per-event list (`swc-open` / `swc-after-open` / `swc-close` with `detail.source` / `swc-after-close`) plus the reduced-motion synchronous-dispatch note, like Tooltip.

> 

Reviewer: Is it worth adding a nested popover example? Or a popover with a nested tooltip?

Author: I have one on the tests stories: https://swcpreviews.z13.web.core.windows.net/pr-6356/docs/gen2-storybook-dev/?path=/story/components-popover-tests--nested-layers-test

I wasn't sure if I should add it to the docs because I'm also not sure what our guidance around nested Popovers is 😅 

Let me know if I should, and I will add it to the docs

Reviewer: ok, good to have the test. we can wait until we have a reasonable use case to demonstrate (also another candidate for a future patterns section!)

## Accessibility

### Features

- The popover needs an **accessible name**: set `accessible-label` and its value is applied to the surface. The name is required because the surface is exposed with `role="dialog"` internally, and the component warns in development when a popover opens without one.
- Opening the popover moves focus into it and closing returns focus to the trigger.
- The resolved trigger receives `aria-expanded`, a durable `aria-controls` relationship wired through the element-reference IDL (so it works across shadow boundaries), and `aria-haspopup="dialog"`. For an SWC component trigger with an open shadow root, the inner `<button>` receives the ARIA.

> 

Reviewer 2: Can you create Keyboard navigation section with list similar to the one in Tabs?

Author: Done. Added a `#### Keyboard navigation` list (Tab/Shift+Tab, Enter/Space, Escape) under Accessibility, matching the Tabs pattern.

### Naming the popover

Set **`accessible-label`** to give the popover an accessible name:

```html
<swc-popover accessible-label="Account settings" for="account-trigger">
  <!-- content -->
</swc-popover>
```

### Best practices

- Give the trigger a clear, visible label; the popover does not name its own trigger.
- Always set `accessible-label`; the surface is a dialog in both modes and needs a name.
- Use `modal` only when the task genuinely needs to block the page; prefer the default mode for menus and contextual content.

> - Ensure a modal popover contains focusable content so focus has somewhere to land.

Reviewer: I would kind of like to demonstrate this, but at minimum it's appropriate to note this for keyboard accessibility.

```suggestion
- Ensure a modal popover contains focusable content so focus has somewhere to land.
- Set `tabindex="0"` on regions that may scroll to ensure keyboard user scroll access.
```

Author: Added!
